pkg/common/log/reopen.go

```diff
@@ -0,0 +1,106 @@
+package log
+
+import (
+	"context"
+	"io"
+	"os"
+	"os/signal"
+	"sync"
+	"syscall"
+)
+
+const (
+	_fileFlags    = os.O_APPEND | os.O_CREATE | os.O_WRONLY
+	_fileMode     = 0640
+	_reopenSignal = syscall.SIGUSR2
+)
+
+var _ ReopenableWriteCloser = (*ReopenableFile)(nil)
+
+// Reopener inspired by https://github.com/client9/reopen
+type Reopener interface {
+	Reopen() error
+}
+
+type ReopenableWriteCloser interface {
+	Reopener
+	io.WriteCloser
+}
+
+type ReopenableFile struct {
+	name string
+	f    *os.File
+	mu   sync.Mutex
+}
+
+func NewReopenableFile(name string) (*ReopenableFile, error) {
+	file, err := os.OpenFile(name, _fileFlags, _fileMode)
+	if err != nil {
+		return nil, err
+	}
+	return &ReopenableFile{
+		name: name,
+		f:    file,
+	}, nil
+}
+
+func (r *ReopenableFile) Reopen() error {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	if r.f != nil {
+		if err := r.f.Close(); err != nil {
+			return err
+		}
+		r.f = nil
+	}
+	newFile, err := os.OpenFile(r.name, _fileFlags, _fileMode)
+	if err != nil {
+		r.f = nil
+		return err
+	}
+	r.f = newFile
+	return nil
+}
+
+func (r *ReopenableFile) Write(p []byte) (n int, err error) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	return r.f.Write(p)
+}
+
+func (r *ReopenableFile) Close() error {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	return r.f.Close()
+}
+
+// Name implements part of os.FileInfo without needing a lock on the
+// underlying file.
+func (r *ReopenableFile) Name() string {
+	return r.name
+}
+
+func ReopenOnSignal(ctx context.Context, rwc ReopenableWriteCloser) {
+	signalCh := make(chan os.Signal, 1)
+	signal.Notify(signalCh, _reopenSignal)
+	reopenOnSignal(ctx, rwc, signalCh)
+}
+
+func reopenOnSignal(
+	ctx context.Context,
+	rwc ReopenableWriteCloser,
+	signalCh chan os.Signal,
+) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-signalCh:
+			// satisfy errcheck
+			_ = rwc.Reopen()
```

Comment:
If we fail to reopen the logger, we can assume that SPIRE will fail to write to the log file in the future. Some other code depending on `ReopenableFile` to be in a functional state will likely panic once it calls `Write()`. This could lead to some confusing and hard to debug errors because we've lost the root cause of why the logger is inoperable.

When we get an error here, I think we would want to preemptively signal the error back to the parent goroutine and trigger a termination of the process.

Reply:
> If we fail to reopen the logger, we can assume that SPIRE will fail to write to the log file in the future.

I was presuming it would continue writing to the old file descriptor (for the now rotated file). I guess that's not a safe assumption - it's just the behavior when the signal handling fails in test. If the reopen fails in production runtime, most likely the filesystem is read-only or otherwise broken in some way.

> When we get an error here, I think we would want to preemptively signal the error back to the parent goroutine and trigger a termination of the process.

I was hoping to avoid plumbing through an `errCh` from `main`.